File: swift/Sources/LongestRunCDF.swift
/*
    djent - A reimplementation of Fourmilab's ent with several improvements.
    
    Copyright (C) 2017  David Johnston
    Translated to Swift 2026

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Contact. David Johnston dj@deadhat.com
*/

import Foundation

/// Return the probability of the longest run of heads being <= n
/// in a sequence of r uniform coin tosses.
///
/// The original C version used MPFR for arbitrary precision.
/// This Swift version uses Double, which is sufficient for typical
/// entropy analysis data sizes. For very large r values, a
/// multi-precision library (e.g. wrapping GMP/MPFR via C interop)
/// would be needed for full accuracy.
///
/// Formula:
///   first  = exp( -(r+1) / (2^(n+1) - n - 2) )
///   second = (2^(n+1) - 1) / (2^(n+1) - (n+2)/2)
///   result = first * second
func longestRunCDF(n: UInt, r: UInt) -> Double {
    let nd = Double(n)
    let rd = Double(r)

    // 2^(n+1)
    let twoToNPlus1 = pow(2.0, nd + 1.0)

    // first = exp( -(r+1) / (2^(n+1) - n - 2) )
    let topa = rd + 1.0
    let bottoma = twoToNPlus1 - nd - 2.0

    // Guard against division by zero for very small n
    guard bottoma != 0.0 else { return 0.0 }

    let first = exp(-topa / bottoma)

    // second = (2^(n+1) - 1) / (2^(n+1) - (n+2)/2)
    let topb = twoToNPlus1 - 1.0
    let nPlus2Over2 = (nd + 2.0) / 2.0
    let bottomb = twoToNPlus1 - nPlus2Over2

    guard bottomb != 0.0 else { return 0.0 }

    let second = topb / bottomb

    return first * second
}

File: zig/src/longest_run_cdf.zig
//! djent - Longest run CDF computation.
//!
//! Copyright (C) 2017  David Johnston
//! Translated to Zig 0.15.x, 2025
//!
//! Contact. David Johnston dj@deadhat.com

const std = @import("std");

/// Return the probability of the longest run of heads being <= n
/// in a sequence of r uniform coin tosses.
///
/// The original C version used MPFR for arbitrary precision.
/// This Zig version uses f64, sufficient for typical entropy analysis data sizes.
pub fn longestRunCDF(n: u32, r: u64) f64 {
    const nd: f64 = @floatFromInt(n);
    const rd: f64 = @floatFromInt(r);

    const two_to_n_plus_1 = std.math.pow(f64, 2.0, nd + 1.0);

    const top_a = rd + 1.0;
    const bottom_a = two_to_n_plus_1 - nd - 2.0;
    if (bottom_a == 0.0) return 0.0;

    const first = @exp(-top_a / bottom_a);

    const top_b = two_to_n_plus_1 - 1.0;
    const n_plus_2_over_2 = (nd + 2.0) / 2.0;
    const bottom_b = two_to_n_plus_1 - n_plus_2_over_2;
    if (bottom_b == 0.0) return 0.0;

    const second = top_b / bottom_b;

    return first * second;
}
